File: boletim2/src/venda.c
#include "headers/venda.h"

#include <ctype.h>
/*eliminar isto, pus só para os printfs para fazer debug*/
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

struct venda{
	Produto produto;
	Preco preco;
	Quantidade quantidade;
	Promo promo;
	Cliente cliente;
	Mes mes;
	Filial filial;
};


/*static Venda nova_venda();*/
static Boolean valida_cliente_venda(String string, CClientes clientes);
static Boolean valida_filial_venda(String string);
static Boolean valida_quantidade_venda(String string);
static Boolean valida_mes_venda(String string);
static Boolean valida_preco_venda(String string);
static Boolean valida_produto_venda(String string, CProdutos produtos);
static Boolean valida_promo_venda(String string);


Cliente get_cliente_venda(Venda venda){
	return venda->cliente;
}

Filial get_filial_venda(Venda venda){
	return venda->filial;
}

Quantidade get_quantidade_venda(Venda venda){
	return venda->quantidade;
}

Mes get_mes_venda(Venda venda){
	return venda->mes;
}

Preco get_preco_venda(Venda venda){
	return venda->preco;
}

Produto get_produto_venda(Venda venda){
	return venda->produto;
}

Promo get_promo_venda(Venda venda){
	return venda->promo;
}

int valida_venda(String string, CClientes clientes, CProdutos produtos){
	
	int i, tamanho;
	String aux, campo;
	aux = malloc(MAXBUFVENDA*sizeof(char));
	char * campos[MAXCAMPOSVENDA] = {NULL};
	i = 0;

	strcpy(aux, string);

	campo = strtok(aux, " \n\r");
	tamanho = strlen(campo) + 1;
	campos[i] = malloc(tamanho*sizeof(char));
	strcpy(campos[i], campo);

	i++; 

	/*MELHORAR ESTA FUNÇÃO: ADICIONAR A CONDICAO "i<MAXCAMPOSVENDA+1" AO 
	WHILE E POR NO PRIMEIRO CHECK campos[MAXCAMPOSVENDA] != NULL EM VEZ 
	DE ARGS != MAXCAMPOSVENDA, ELIMINAR O ARGS PORQUE NAO FAZ FALTA*/
	/*coloca num array de strings várias strings correspondente a uma venda
	fragmentada nos seus vários campos*/
	while(i < MAXCAMPOSVENDA){
		campo = strtok(NULL, " \n\r");
		tamanho = strlen(campo) + 1;
		campos[i] = malloc(tamanho*sizeof(char));
		strcpy(campos[i], campo);
		i++; 
	}

	campo = strtok(NULL, " \n\r");

	if(campo != NULL)
		return FALSE;

	if(!valida_produto_venda(campos[0], produtos))
		return FALSE;

	if(!valida_preco_venda(campos[1]))
		return FALSE;

	if(!valida_quantidade_venda(campos[2]))
		return FALSE;

	if(!valida_promo_venda(campos[3]))
		return FALSE;

	if(!(valida_cliente_venda(campos[4], clientes)))
		return FALSE;

	if(!(valida_mes_venda(campos[5])))
		return FALSE;

	if(!(valida_filial_venda(campos[6])))
		return FALSE;

	return TRUE;
}

/*static Venda nova_venda(){
	return NULL;
}*/

static Boolean valida_cliente_venda(String string, CClientes clientes){
	if(valida_cliente(string) == FALSE)
		return FALSE;
	if(existe_cliente(string, clientes) == FALSE)
		return FALSE; 
	return TRUE;
}

static Boolean valida_filial_venda(String string){
	int i;
	for(i = 0; i < ((int) strlen(string)); i++)
		if(!isdigit(string[i]))
			return FALSE;
	if (atoi(string) < 1 || atoi(string) > 3)
		return FALSE;
	return TRUE;
}

static Boolean valida_quantidade_venda(String string){
	int i;
	for(i = 0; i < ((int) strlen(string)); i++)
		if(!isdigit(string[i]))
			return FALSE;
	if(atoi(string) < 1 || atoi(string) > 200)
		return FALSE;	
	return TRUE;
}

static Boolean valida_mes_venda(String string){
	int i;
	for(i = 0; i < ((int) strlen(string)); i++)
		if(!isdigit(string[i]))
			return FALSE;
	if (atoi(string) < 1 || atoi(string) > 12)
		return FALSE;
	return TRUE;
}

static Boolean valida_preco_venda(String string){
	int i;
	for(i = 0; i < ((int) strlen(string)) && string[i]!='.'; i++)
		if(!isdigit(string[i]))
			return FALSE;
	for(i=i+1; i < ((int) strlen(string)); i++)
		if(!isdigit(string[i]))
			return FALSE;
	if(atof(string) > 9999.99 || atof(string) < 0 )
		return FALSE; 
	return TRUE;
}

static Boolean valida_produto_venda(String string, CProdutos produtos){
	if(valida_produto(string) == FALSE)
		return FALSE;
	if(existe_produto(string, produtos) == FALSE)
		return FALSE; 
	return TRUE;
}

static Boolean valida_promo_venda(String string){
	if(!(strcmp(string, "N") == 0 || strcmp(string, "P") == 0))
		return FALSE;
	return TRUE;
}

File: projeto_java/vendum/src/Venda.java
import java.io.Serializable;

public class Venda implements Serializable{
    private String produto;
    private double preco;
    private int quantidade;
    private char promo;
    private String cliente;
    private int mes;
    private int filial;

    /**
     * Construtor vazio.
     */
    public Venda(){
        this.produto = null;
        this.preco = 0;
        this.quantidade = 0;
        this.promo = '0';
        this.cliente = null;
        this.mes = 0;
        this.filial = 0;
    }

    /**
     * Construtor parametrizado.
     * @param produto
     * @param preco
     * @param quantidade
     * @param promo
     * @param cliente
     * @param mes
     * @param filial
     */
    public Venda(String produto, double preco, int quantidade, char promo, String cliente, int mes, int filial) {
        this.produto = produto;
        this.preco = preco;
        this.quantidade = quantidade;
        this.promo = promo;
        this.cliente = cliente;
        this.mes = mes;
        this.filial = filial;
    }

    public String getProduto() {
        return produto;
    }

    public void setProduto(String produto) {
        this.produto = produto;
    }

    public double getPreco() {
        return preco;
    }

    public void setPreco(double preco) {
        this.preco = preco;
    }

    public int getQuantidade() {
        return quantidade;
    }

    public void setQuantidade(int quantidade) {
        this.quantidade = quantidade;
    }

    public char getPromo() {
        return promo;
    }

    public void setPromo(char promo) {
        this.promo = promo;
    }

    public String getCliente() {
        return cliente;
    }

    public void setCliente(String cliente) {
        this.cliente = cliente;
    }

    public int getMes() {
        return mes;
    }

    public void setMes(int mes) {
        this.mes = mes;
    }

    public int getFilial() {
        return filial;
    }

    public void setFilial(int filial) {
        this.filial = filial;
    }

    public double getValor(){
        return preco*quantidade;
    }

    /**
     * Método que faz o parsing e que valida uma venda.
     * @param cclientes
     * @param cprodutos
     * @param linha
     * @return null se a venda em questão for inválida ou uma instância de venda se esta for válida.
     */
    public Venda venda_valida(CClientes cclientes, CProdutos cprodutos, String linha){
        String[] campos = linha.split(" ");

        if(campos.length != Globais.NRCAMPOS)
            return null;

        if(!valida_produto_venda(campos[0], cprodutos))
            return null;

        if(!valida_preco_venda(campos[1]))
            return null;

        if(!valida_quantidade_venda(campos[2]))
            return null;

        if(!valida_promo_venda(campos[3]))
            return null;

        if(!valida_cliente_venda(campos[4], cclientes))
            return null;

        if(!valida_mes_venda(campos[5]))
            return null;

        if(!valida_filial_venda(campos[6]))
            return null;

        Venda venda = new Venda(campos[0], Double.parseDouble(campos[1]), Integer.parseInt(campos[2]), campos[3].charAt(0), campos[4], Integer.parseInt(campos[5]), Integer.parseInt(campos[6]));

        return venda;
    }

    public boolean valida_filial_venda(String string){
        int i;
        for(i = 0; i < string.length(); i++)
            if(!Character.isDigit(string.charAt(i)))
                return false;
        if (Integer.parseInt(string) < 1 || Integer.parseInt(string) > 3)
            return false;
        return true;
    }

    public boolean valida_mes_venda(String string){
        int i;
        for(i = 0; i < string.length(); i++)
            if(!Character.isDigit(string.charAt(i)))
                return false;
        if (Integer.parseInt(string) < 1 || Integer.parseInt(string) > 12)
            return false;
        return true;
    }

    private boolean valida_cliente_venda(String string, CClientes clientes){
        if(!clientes.valida_cliente(string))
            return false;
        if(!clientes.existe_cliente(string))
            return false;
        return true;
    }

    private boolean valida_promo_venda(String string){
        if(string.equals("N") || string.equals("P"))
            return true;
        return false;
    }

    private boolean valida_quantidade_venda(String string){
        int i;
        for(i = 0; i < string.length(); i++)
            if(!Character.isDigit(string.charAt(i)))
                return false;
        if(Integer.parseInt(string) < 1 || Integer.parseInt(string) > 200)
            return false;
        return true;
    }

    private boolean valida_preco_venda(String string){
        int i;
        for(i = 0; i < string.length() && string.charAt(i)!='.'; i++)
            if(!Character.isDigit(string.charAt(i)))
                return false;
        for(i=i+1; i < string.length(); i++)
            if(!Character.isDigit(string.charAt(i)))
                return false;
        if(Double.parseDouble(string) > 9999.99 || Double.parseDouble(string) < 0 )
            return false;
        return true;
    }

    private boolean valida_produto_venda(String produto, CProdutos produtos){
        if(!produtos.valida_produto(produto))
            return false;
        if(!produtos.existe_produto(produto))
            return false;
        return true;
    }
}
